File: app/src/main/java/closer/vlllage/com/closer/GroupActivity.java
package closer.vlllage.com.closer;

import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v7.widget.LinearLayoutManager;
import android.view.View;

import java.util.List;

import closer.vlllage.com.closer.handler.data.AccountHandler;
import closer.vlllage.com.closer.handler.data.ApiHandler;
import closer.vlllage.com.closer.handler.data.PermissionHandler;
import closer.vlllage.com.closer.handler.data.PersistenceHandler;
import closer.vlllage.com.closer.handler.data.RefreshHandler;
import closer.vlllage.com.closer.handler.event.EventDetailsHandler;
import closer.vlllage.com.closer.handler.group.GroupActionHandler;
import closer.vlllage.com.closer.handler.group.GroupActivityTransitionHandler;
import closer.vlllage.com.closer.handler.group.GroupContactsHandler;
import closer.vlllage.com.closer.handler.group.GroupHandler;
import closer.vlllage.com.closer.handler.group.GroupMemberHandler;
import closer.vlllage.com.closer.handler.group.GroupMessageAttachmentHandler;
import closer.vlllage.com.closer.handler.group.GroupMessageMentionHandler;
import closer.vlllage.com.closer.handler.group.GroupMessagesHandler;
import closer.vlllage.com.closer.handler.group.GroupToolbarHandler;
import closer.vlllage.com.closer.handler.group.GroupViewHolder;
import closer.vlllage.com.closer.handler.group.PinnedMessagesHandler;
import closer.vlllage.com.closer.handler.group.SearchGroupsAdapter;
import closer.vlllage.com.closer.handler.helpers.ActivityHandler;
import closer.vlllage.com.closer.handler.helpers.AlertHandler;
import closer.vlllage.com.closer.handler.helpers.ApplicationHandler;
import closer.vlllage.com.closer.handler.helpers.ConnectionErrorHandler;
import closer.vlllage.com.closer.handler.helpers.DefaultAlerts;
import closer.vlllage.com.closer.handler.helpers.DisposableHandler;
import closer.vlllage.com.closer.handler.helpers.GroupColorHandler;
import closer.vlllage.com.closer.handler.helpers.GroupScopeHandler;
import closer.vlllage.com.closer.handler.helpers.KeyboardHandler;
import closer.vlllage.com.closer.handler.helpers.MiniWindowHandler;
import closer.vlllage.com.closer.handler.helpers.ResourcesHandler;
import closer.vlllage.com.closer.handler.helpers.SortHandler;
import closer.vlllage.com.closer.handler.helpers.SystemSettingsHandler;
import closer.vlllage.com.closer.handler.helpers.TimerHandler;
import closer.vlllage.com.closer.handler.helpers.TopHandler;
import closer.vlllage.com.closer.store.StoreHandler;
import closer.vlllage.com.closer.store.models.Group;
import closer.vlllage.com.closer.store.models.GroupMember;
import closer.vlllage.com.closer.store.models.GroupMember_;
import closer.vlllage.com.closer.store.models.Group_;
import closer.vlllage.com.closer.ui.CircularRevealActivity;
import io.objectbox.android.AndroidScheduler;
import io.objectbox.query.QueryBuilder;

import static android.Manifest.permission.READ_CONTACTS;

public class GroupActivity extends CircularRevealActivity {

    public static final String EXTRA_GROUP_ID = "groupId";
    public static final String EXTRA_RESPOND = "respond";

    public GroupViewHolder view;
    public String groupId;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_group);
        view = new GroupViewHolder(findViewById(android.R.id.content));

        findViewById(R.id.closeButton).setOnClickListener(view -> finish());

        $(ApiHandler.class).setAuthorization($(AccountHandler.class).getPhone());

        $(TimerHandler.class).postDisposable(() -> $(RefreshHandler.class).refreshAll(), 1625);

        $(GroupToolbarHandler.class).attach(findViewById(R.id.eventToolbar));

        $(GroupHandler.class).attach(view.groupName, view.backgroundPhoto, view.groupAbout, view.peopleInGroup, findViewById(R.id.settingsButton));
        handleIntent(getIntent());

        view.groupAbout.setOnClickListener(v -> {
            if ($(GroupHandler.class).getGroup() == null || $(GroupHandler.class).getGroup().getAbout() == null) {
                return;
            }

            $(DefaultAlerts.class).message(
                    $(ResourcesHandler.class).getResources().getString(R.string.about_this_group),
                    $(GroupHandler.class).getGroup().getAbout()
            );
        });

        $(GroupActionHandler.class).attach(view.actionFrameLayout, findViewById(R.id.actionRecyclerView));
        $(GroupMessagesHandler.class).attach(view.messagesRecyclerView, view.replyMessage, view.sendButton, view.sendMoreButton, findViewById(R.id.sendMoreLayout));
        $(PinnedMessagesHandler.class).attach(view.pinnedMessagesRecyclerView);
        $(GroupMessageMentionHandler.class).attach(view.mentionSuggestionsLayout, findViewById(R.id.mentionSuggestionRecyclerView), mention -> {
            $(GroupMessagesHandler.class).insertMention(mention);
        });
        $(MiniWindowHandler.class).attach(view.groupName, findViewById(R.id.backgroundColor), this::finish);

        findViewById(R.id.settingsButton).setOnClickListener(view -> {
            $(GroupMemberHandler.class).changeGroupSettings($(GroupHandler.class).getGroup());
        });

        if ($(PersistenceHandler.class).getPhoneId() != null) {
            $(DisposableHandler.class).add($(StoreHandler.class).getStore().box(GroupMember.class).query()
                    .equal(GroupMember_.group, groupId)
                    .equal(GroupMember_.phone, $(PersistenceHandler.class).getPhoneId())
                    .build().subscribe().on(AndroidScheduler.mainThread()).observer(groupMembers -> {
                        GroupMember groupMember = groupMembers.isEmpty() ? null : groupMembers.get(0);

                        if (groupMember == null) {
                            groupMember = new GroupMember();
                        }

                        if (groupMember.isMuted()) {
                            findViewById(R.id.notificationSettingsButton).setOnClickListener(view -> $(GroupMemberHandler.class).changeGroupSettings($(GroupHandler.class).getGroup()));
                            findViewById(R.id.notificationSettingsButton).setVisibility(View.VISIBLE);
                        } else {
                            findViewById(R.id.notificationSettingsButton).setVisibility(View.GONE);
                        }
                    }));
        }

        view.replyMessage.setOnClickListener(view -> {
            $(GroupActionHandler.class).show(false);
        });

        $(DisposableHandler.class).add($(GroupHandler.class).onGroupUpdated().subscribe(group -> {
            $(PinnedMessagesHandler.class).show(group);
            $(GroupHandler.class).setGroupBackground(group);
        }));

        $(DisposableHandler.class).add($(GroupHandler.class).onGroupChanged().subscribe(group -> {
            $(PinnedMessagesHandler.class).show(group);

            findViewById(R.id.backgroundColor).setBackgroundResource($(GroupColorHandler.class).getColorBackground(group));

            $(GroupScopeHandler.class).setup(group, view.scopeIndicatorButton);

            view.peopleInGroup.setSelected(true);
            view.peopleInGroup.setOnClickListener(view -> toggleContactsView());

            $(GroupContactsHandler.class).attach(group, view.contactsRecyclerView, view.searchContacts, view.showPhoneContactsButton);

            view.showPhoneContactsButton.setOnClickListener(v -> {
                if ($(PermissionHandler.class).denied(READ_CONTACTS)) {
                    $(AlertHandler.class).make()
                            .setTitle($(ResourcesHandler.class).getResources().getString(R.string.enable_contacts_permission))
                            .setMessage($(ResourcesHandler.class).getResources().getString(R.string.enable_contacts_permission_rationale))
                            .setPositiveButton($(ResourcesHandler.class).getResources().getString(R.string.open_settings))
                            .setPositiveButtonCallback(alertResult -> $(SystemSettingsHandler.class).showSystemSettings())
                            .show();
                    return;
                }

                $(PermissionHandler.class).check(READ_CONTACTS).when(granted -> {
                    if (granted) {
                        $(GroupContactsHandler.class).showContactsForQuery();
                        view.showPhoneContactsButton.setVisibility(View.GONE);
                    }
                });
            });

            $(GroupHandler.class).setGroupBackground(group);
        }, error -> $(ConnectionErrorHandler.class).notifyConnectionError()));

        $(DisposableHandler.class).add($(GroupHandler.class).onEventChanged().subscribe(event -> {
            view.groupDetails.setVisibility(View.VISIBLE);
            view.groupDetails.setText($(EventDetailsHandler.class).formatEventDetails(event));
        }, error -> $(ConnectionErrorHandler.class).notifyConnectionError()));

        view.shareWithRecyclerView.setLayoutManager(new LinearLayoutManager(
                view.shareWithRecyclerView.getContext(),
                LinearLayoutManager.VERTICAL,
                false
        ));

        $(DisposableHandler.class).add($(GroupToolbarHandler.class).getIsShareActiveObservable()
                .subscribe(isShareActive -> {
                    showMessagesView(!isShareActive);

                    QueryBuilder<Group> queryBuilder = $(StoreHandler.class).getStore().box(Group.class).query();
                    List<Group> groups = queryBuilder.sort($(SortHandler.class).sortGroups()).notEqual(Group_.physical, true).build().find();

                    SearchGroupsAdapter searchGroupsAdapter = new SearchGroupsAdapter($(GroupHandler.class), (group, view) -> {
                        boolean success = $(GroupMessageAttachmentHandler.class).shareEvent($(GroupHandler.class).onEventChanged().getValue(), group);

                        if (success) {
                            ((CircularRevealActivity) $(ActivityHandler.class).getActivity())
                                    .finish(() -> $(GroupActivityTransitionHandler.class).showGroupMessages(view, group.getId()));
                        } else {
                            $(DefaultAlerts.class).thatDidntWork();
                        }
                    }, null);

                    searchGroupsAdapter.setGroups(groups);
                    searchGroupsAdapter.setActionText($(ResourcesHandler.class).getResources().getString(R.string.share));
                    searchGroupsAdapter.setIsSmall(true);

                    view.shareWithRecyclerView.setAdapter(searchGroupsAdapter);
                }));
    }

    @Override
    protected void onNewIntent(Intent intent) {
        setIntent(intent);
        handleIntent(intent);
        setSourceBounds(intent.getSourceBounds());
        reveal();
    }

    private void handleIntent(Intent intent) {
        if (intent != null && intent.hasExtra(EXTRA_GROUP_ID)) {
            groupId = intent.getStringExtra(EXTRA_GROUP_ID);
            $(GroupHandler.class).setGroupById(groupId);

            if (intent.hasExtra(EXTRA_RESPOND)) {
                view.replyMessage.postDelayed(() -> {
                    view.replyMessage.requestFocus();
                    $(KeyboardHandler.class).showKeyboard(view.replyMessage, true);
                }, 500);
            }
        }
    }

    @Override
    protected void onResume() {
        super.onResume();
        $(ApplicationHandler.class).getApp().$(TopHandler.class).setGroupActive(groupId);
    }

    @Override
    protected void onPause() {
        super.onPause();
        $(ApplicationHandler.class).getApp().$(TopHandler.class).setGroupActive(null);
    }

    @Override
    protected int getBackgroundId() {
        return R.id.background;
    }

    private void toggleContactsView() {
        $(GroupToolbarHandler.class).getIsShareActiveObservable().onNext(false);
        showMessagesView(view.replyMessage.getVisibility() == View.GONE);
        view.shareWithRecyclerView.setVisibility(View.GONE);
    }

    private void showMessagesView(boolean show) {
        $(GroupMessagesHandler.class).showSendMoreOptions(false);

        if (show) {
            view.shareWithRecyclerView.setVisibility(View.GONE);
            view.messagesLayoutGroup.setVisibility(View.VISIBLE);
            view.membersLayoutGroup.setVisibility(View.GONE);

            if (view.replyMessage.getText().toString().isEmpty()) {
                view.sendMoreButton.setVisibility(View.VISIBLE);
            } else {
                view.sendMoreButton.setVisibility(View.GONE);
            }

            view.showPhoneContactsButton.setVisibility(View.GONE);
        } else {
            view.shareWithRecyclerView.setVisibility(View.VISIBLE);
            view.messagesLayoutGroup.setVisibility(View.GONE);
            view.membersLayoutGroup.setVisibility(View.VISIBLE);

            view.sendMoreButton.setVisibility(View.GONE);

            $(GroupActionHandler.class).cancelPendingAnimation();

            if(!$(PermissionHandler.class).has(READ_CONTACTS)) {
                view.showPhoneContactsButton.setVisibility(View.VISIBLE);
            }

            if($(PermissionHandler.class).has(READ_CONTACTS)) {
                $(GroupContactsHandler.class).showContactsForQuery();
            }

            view.searchContacts.setText("");
        }
    }
}

File: app/src/main/java/closer/vlllage/com/closer/GroupActivity.kt
package closer.vlllage.com.closer

import android.animation.ObjectAnimator
import android.content.Intent
import android.os.Bundle
import android.view.Gravity
import androidx.constraintlayout.widget.ConstraintLayout
import androidx.fragment.app.Fragment
import closer.vlllage.com.closer.databinding.ActivityGroupBinding
import closer.vlllage.com.closer.extensions.visible
import closer.vlllage.com.closer.handler.FeatureHandler
import closer.vlllage.com.closer.handler.FeatureType
import closer.vlllage.com.closer.handler.data.ApiHandler
import closer.vlllage.com.closer.handler.data.DataHandler
import closer.vlllage.com.closer.handler.data.PersistenceHandler
import closer.vlllage.com.closer.handler.data.RefreshHandler
import closer.vlllage.com.closer.handler.event.EventDetailsHandler
import closer.vlllage.com.closer.handler.group.*
import closer.vlllage.com.closer.handler.helpers.*
import closer.vlllage.com.closer.handler.map.MeetHandler
import closer.vlllage.com.closer.handler.phone.NameHandler
import closer.vlllage.com.closer.handler.settings.SettingsHandler
import closer.vlllage.com.closer.handler.settings.UserLocalSetting
import closer.vlllage.com.closer.store.models.Group
import closer.vlllage.com.closer.store.models.Phone
import closer.vlllage.com.closer.ui.CircularRevealActivity
import com.bumptech.glide.load.resource.drawable.DrawableTransitionOptions
import io.reactivex.rxjava3.android.schedulers.AndroidSchedulers
import java.text.SimpleDateFormat
import java.util.*

class GroupActivity : CircularRevealActivity() {

    lateinit var groupId: String

    private lateinit var binding: ActivityGroupBinding

    private var initialContent: ContentViewType? = null

    private var restoreOption = false

    private var contentView: ContentViewType = ContentViewType.MESSAGES
        set(value) {
            field = value
            setContent(when (field) {
                ContentViewType.MESSAGES -> GroupMessagesFragment()
                ContentViewType.CONTACTS -> GroupContactsFragment()
                ContentViewType.EVENTS -> GroupEventsFragment()
                ContentViewType.REVIEWS -> GroupReviewsFragment()
                ContentViewType.GROUP_PHOTOS -> GroupPhotosFragment()
                ContentViewType.GROUP_ABOUT -> GroupAboutFragment()
                ContentViewType.PHONE_MESSAGES -> PhoneMessagesFragment()
                ContentViewType.PHONE_PHOTOS -> PhonePhotosFragment()
                ContentViewType.PHONE_GROUPS -> PhoneGroupsFragment()
                ContentViewType.PHONE_ABOUT -> PhoneAboutFragment()
                else -> Fragment()
            })
        }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityGroupBinding.inflate(layoutInflater).also {
            setContentView(it.root)
        }

        bindViewEvents()
        bindToGroup()

        if (on<FeatureHandler>().has(FeatureType.FEATURE_MANAGE_PUBLIC_GROUP_SETTINGS)) {
            binding.settingsButton.visible = true
        }

        handleIntent(intent)

        on<GroupToolbarHandler>().attach(binding.eventToolbar) {
            if ((binding.profilePhoto.layoutParams as ConstraintLayout.LayoutParams).matchConstraintPercentHeight == 1f) {
                return@attach
            }

            val initialHeight = binding.profilePhoto.measuredHeight
            val finalHeight = 0 //on<ResourcesHandler>().resources.getDimensionPixelSize(R.dimen.profilePhotoCollapsedHeight)

            ObjectAnimator.ofFloat(0f, 1f).apply {
                duration = 150
                addUpdateListener {
                    val params = binding.profilePhoto.layoutParams as ConstraintLayout.LayoutParams
                    params.apply {
                        height = (initialHeight + it.animatedFraction * (finalHeight - initialHeight)).toInt()
                        matchConstraintPercentHeight = if (height == 0) 0f else 1f
                        binding.profilePhoto.layoutParams = this
                    }
                }
                start()
            }
        }

        on<DisposableHandler>().add(on<GroupToolbarHandler>().contentView
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe { contentView = it })

        on<MiniWindowHandler>().attach(binding.groupName, binding.backgroundColor) { finish() }

        on<DisposableHandler>().add(on<LightDarkHandler>().onLightChanged.subscribe {
            binding.closeButton.imageTintList = it.tint
            binding.scopeIndicatorButton.imageTintList = it.tint
            binding.settingsButton.imageTintList = it.tint
            binding.notificationSettingsButton.imageTintList = it.tint
            binding.groupName.setTextColor(it.text)
            binding.groupAbout.setTextColor(it.text)
            binding.groupDetails.setTextColor(it.text)
            binding.groupRatingCount.setTextColor(it.text)
            binding.peopleInGroup.setTextColor(it.text)
            binding.backgroundPhoto.alpha = if (it.light) .15f else 1f
        })
    }

    private fun bindToGroup() {
        on<GroupHandler> {
            if (on<SettingsHandler>().get(UserLocalSetting.CLOSER_SETTINGS_USE_LIGHT_THEME)) {
                on<LightDarkHandler>().setLight(true)
            }

            onGroupNotFound {
               finish()
            }

            onGroupMemberChanged { groupMember ->
                binding.notificationSettingsButton.visible = groupMember.muted
            }

            onGroupUpdated { group ->
                setGroupBackground(group)
                setGroupRating(group)
                setGroupAbout(group)
            }

            onContactInfoChanged { redrawContacts(it) }

            onGroupChanged { group ->
                binding.settingsButton.setOnLongClickListener(null)

                on<GroupToolbarHandler>().contentView.onNext(initialContent ?: ContentViewType.MESSAGES)

                on<GroupScopeHandler>().setup(group, binding.scopeIndicatorButton)

                binding.peopleInGroup.isSelected = true

                if (!group.hasPhone()) {
                    setGroupProfilePhoto(null)
                }

                setGroupBackground(group)
                setGroupRating(group)

                showGroupName(group)

                if (!group.hasEvent()) {
                    binding.groupDetails.visible = false
                    binding.groupDetails.text = ""
                }

                binding.peopleInGroup.text = ""

                setGroupAbout(group)

                on<GroupScopeHandler>().setup(group, binding.scopeIndicatorButton)

                binding.peopleInGroup.isSelected = true

                setGroupBackground(group)

                on<ApplicationHandler>().app.on<TopHandler>().setGroupActive(group.id!!)

                if (restoreOption) {
                    binding.meetLayout.visible = true
                    binding.meetFalse.visible = false
                    binding.meetPrompt.gravity = Gravity.START
                    binding.meetPrompt.setPaddingRelative(
                            on<ResourcesHandler>().resources.getDimensionPixelSize(R.dimen.padDialog),
                            binding.meetPrompt.paddingTop,
                            binding.meetPrompt.paddingEnd,
                            binding.meetPrompt.paddingBottom,
                    )
                    binding.meetTrue.setText(R.string.restore)
                    binding.meetTrue.setCompoundDrawablesRelativeWithIntrinsicBounds(0, 0, R.drawable.ic_history_black_24dp, 0)

                    binding.meetPrompt.text = on<ResourcesHandler>().resources.getString(R.string.group_expired_on, on<TimeStr>().approx(group.updated?.let {
                        Calendar.getInstance(TimeZone.getDefault()).let { calendar ->
                            calendar.time = it

                            calendar.add(Calendar.MONTH, 3)

                            calendar.time
                        }
                    }, true))
                    binding.meetTrue.setOnClickListener {
                        on<GroupMessageAttachmentHandler>().postGroupEvent(groupId, "@${on<PersistenceHandler>().phoneId} breathed new life into the group!")
                        binding.meetLayout.visible = false
                    }
                }

                binding.groupDetails.setOnClickListener(null)
            }

            onEventChanged { event ->
                binding.groupDetails.visible = true
                binding.groupDetails.text = on<EventDetailsHandler>().formatEventDetails(event)
                binding.groupDetails.setOnClickListener {
                    val timeFormatter = SimpleDateFormat("MMMM d, yyyy${if (event.allDay) "" else " h:mma"}", Locale.US)
                    on<DefaultAlerts>().message("${timeFormatter.format(event.startsAt!!)} to ${timeFormatter.format(event.endsAt!!)}")
                }
            }

            onPhoneUpdated { phone ->
                setGroupProfilePhoto(phone.photo)
                binding.groupDetails.visible = false
                binding.groupDetails.text = ""
                binding.groupAbout.visible = !phone.status.isNullOrBlank()
                binding.groupAbout.text = phone.status ?: ""

                binding.settingsButton.setOnLongClickListener {
                    on<AlertHandler>().make().apply {
                        title = "Terms of Use"
                        negativeButton = "Bad"
                        positiveButton = "Good"
                        negativeButtonCallback = {
                            updateTerms(phone, false)
                        }
                        positiveButtonCallback = {
                            updateTerms(phone, true)
                        }
                        show()
                    }

                    true
                }

                if (on<MatchHandler>().active) {
                    binding.meetLayout.visible = true
                    binding.meetFalse.visible = true
                    binding.meetTrue.visible = true
                    binding.meetPrompt.gravity = Gravity.CENTER
                    binding.meetPrompt.text = on<ResourcesHandler>().resources.getString(R.string.want_to_meet_phone, on<NameHandler>().getName(phone))
                    binding.meetTrue.setOnClickListener {
                        on<MeetHandler>().meet(phone.id!!, true)
                        if (!on<MeetHandler>().next()) finish()
                    }
                    binding.meetFalse.setOnClickListener {
                        on<MeetHandler>().meet(phone.id!!, false)
                        if (!on<MeetHandler>().next()) finish()
                    }
                }
            }
        }
    }

    private fun updateTerms(phone: Phone, good: Boolean) {
        on<ApiHandler>().terms(phone.id!!, good)
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe({
                    if (it.success) {
                        on<DefaultAlerts>().message("Terms updated for ${on<NameHandler>().getName(phone)}")
                    } else {
                        on<DefaultAlerts>().thatDidntWork()
                    }
                }, {
                    on<DefaultAlerts>().thatDidntWork()
                }).also {
                    on<DisposableHandler>().add(it)
                }
    }

    private fun redrawContacts(contactInfo: ContactInfo) {
        val names = mutableListOf<String>()

        names.addAll(contactInfo.contactNames)
        names.addAll(contactInfo.contactInvites)

        if (names.isEmpty()) {
            binding.peopleInGroup.visible = false
            binding.peopleInGroup.setText(R.string.add_contact)
            return
        }

        binding.peopleInGroup.visible = true
        binding.peopleInGroup.text = names.joinToString()
    }

    private fun showGroupName(group: Group?) {
        if (group == null) {
            binding.groupName.setText(R.string.not_found)
            return
        }

        if (group.hasPhone()) {
            on<DisposableHandler>().add(on<DataHandler>().getPhone(group.phoneId!!).subscribe(
                    { binding.groupName.text = on<NameHandler>().getName(it) }, { on<DefaultAlerts>().thatDidntWork() }
            ))
        } else {
            on<GroupNameHelper>().loadName(group, binding.groupName) { it }
        }
    }

    private fun setGroupRating(group: Group) {
        val visible = group.ratingAverage != null && group.ratingCount != null

        binding.groupRatingAverage.visible = visible
        binding.groupRatingCount.visible = visible

        if (!visible) {
            return
        }

        binding.groupRatingAverage.rating = group.ratingAverage!!.toFloat()
        binding.groupRatingCount.text = on<ResourcesHandler>().resources.getQuantityString(R.plurals.review_count_parenthesized, group.ratingCount!!, group.ratingCount)
    }

    private fun setGroupAbout(group: Group) {
        binding.groupAbout.visible = !group.about.isNullOrBlank()
        binding.groupAbout.text = group.about
    }

    private fun setGroupBackground(group: Group) {
        if (group.photo != null) {
            binding.backgroundPhoto.visible = true
            on<ImageHandler>().get().load(group.photo + "?s=512")
                    .transition(DrawableTransitionOptions.withCrossFade())
                    .into(binding.backgroundPhoto)
            binding.backgroundPhoto.setOnClickListener { on<PhotoActivityTransitionHandler>().show(binding.backgroundPhoto, group.photo!!) }

        } else {
            binding.backgroundPhoto.visible = false
        }

        if (on<LightDarkHandler>().isLight()) {
            binding.backgroundColor.setBackgroundResource(R.drawable.color_white_rounded)
        } else {
            binding.backgroundColor.setBackgroundResource(on<GroupColorHandler>().getColorBackground(group))
        }
    }

    private fun setGroupProfilePhoto(photo: String?) {
        if (photo != null) {
            binding.profilePhoto.visible = true
            on<ImageHandler>().get().load("$photo?s=512")
                    .transition(DrawableTransitionOptions.withCrossFade())
                    .into(binding.profilePhoto)
            binding.profilePhoto.setOnClickListener { on<PhotoActivityTransitionHandler>().show(binding.profilePhoto, photo) }

        } else {
            binding.profilePhoto.visible = false
        }
    }

    private fun bindViewEvents() {
        binding.closeButton.setOnClickListener { finish() }

        binding.groupAbout.setOnClickListener {
            on<DefaultAlerts>().message(
                on<GroupHandler>().phone?.let {
                    if (it.id == on<PersistenceHandler>().phoneId) on<ResourcesHandler>().resources.getString(R.string.public_status) else
                        on<ResourcesHandler>().resources.getString(
                            R.string.person_s_status,
                            on<NameHandler>().getName(it)
                        )

                } ?: on<ResourcesHandler>().resources.getString(R.string.about_this_group),
                on<GroupHandler>().phone?.status ?: on<GroupHandler>().group?.about ?: ""
            )
        }

        binding.peopleInGroup.setOnClickListener { toggleContactsView() }

        binding.settingsButton.setOnClickListener {
            on<GroupMemberHandler>().changeGroupSettings(on<GroupHandler>().group)
        }

        binding.notificationSettingsButton.setOnClickListener {
            on<GroupMemberHandler>().mute(on<GroupHandler>().group, false)
        }

        binding.closeButton.visible = !on<SettingsHandler>()[UserLocalSetting.CLOSER_SETTINGS_HIDE_CLOSE_BUTTON]
    }

    override fun onNewIntent(intent: Intent) {
        super.onNewIntent(intent)
        setIntent(intent)
        handleIntent(intent)
        setSourceBounds(intent.sourceBounds)
        reveal()
    }

    private fun handleIntent(intent: Intent?) {
        intent ?: return

        if (intent.hasExtra(EXTRA_GROUP_ID)) {
            groupId = intent.getStringExtra(EXTRA_GROUP_ID)!!
            on<GroupHandler>().setGroupById(groupId)

            if (intent.hasExtra(EXTRA_RESPOND)) {
                on<GroupMessagesHandler>().setIsRespond()
            }

            if (intent.hasExtra(EXTRA_CONTENT)) {
                initialContent = ContentViewType.valueOf(intent.getStringExtra(EXTRA_CONTENT)!!)
            }

            if (intent.hasExtra(EXTRA_NEW_MEMBER)) {
                val disposableGroup = on<DisposableHandler>().group()
                on<GroupHandler>().onGroupChanged(disposableGroup) {
                    disposableGroup.dispose()
                    on<DefaultAlerts>().message(on<ResourcesHandler>().resources.getString(R.string.youre_a_member, it.name ?: on<ResourcesHandler>().resources.getString(R.string.generic_group)))
                }
            }
        } else if (intent.hasExtra(EXTRA_PHONE_ID)) {
            val phoneId = intent.getStringExtra(EXTRA_PHONE_ID)!!
            on<RefreshHandler>().refreshGroupForPhone(phoneId)
            on<DisposableHandler>().add(on<DataHandler>().getGroupForPhone(phoneId).subscribe({
                groupId = it.id!!
                on<GroupHandler>().setGroupById(groupId)
            }, {
                on<DefaultAlerts>().thatDidntWork()
            }))
        }

        if (intent.hasExtra(EXTRA_MEET)) {
            on<MatchHandler>().activate()
        }

        if (intent.hasExtra(EXTRA_RESTORE)) {
            restoreOption = true
        }
    }

    override fun onResume() {
        super.onResume()
        if (this::groupId.isInitialized) {
            on<ApplicationHandler>().app.on<TopHandler>().setGroupActive(groupId)
        }
    }

    override fun onPause() {
        super.onPause()
        on<ApplicationHandler>().app.on<TopHandler>().setGroupActive(null)
    }

    override val backgroundId = R.id.background

    private fun toggleContactsView() {
        on<GroupToolbarHandler>().contentView.onNext(if (on<GroupToolbarHandler>().contentView.value == ContentViewType.CONTACTS) ContentViewType.MESSAGES else ContentViewType.CONTACTS)
    }

    private fun setContent(fragment: Fragment) {
        supportFragmentManager.beginTransaction()
                .replace(R.id.contentFrame, fragment)
                .commitAllowingStateLoss()

    }

    companion object {
        const val EXTRA_GROUP_ID = "groupId"
        const val EXTRA_PHONE_ID = "phoneId"
        const val EXTRA_RESPOND = "respond"
        const val EXTRA_MEET = "meet"
        const val EXTRA_NEW_MEMBER = "newMember"
        const val EXTRA_CONTENT = "content"
        const val EXTRA_RESTORE = "restore"
    }
}
